cpp/oneapi/dal/backend/primitives/rng/host_engine.hpp

```diff
@@ -0,0 +1,124 @@
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <daal/include/algorithms/engines/mt2203/mt2203.h>
+#include <daal/include/algorithms/engines/mcg59/mcg59.h>
+#include <daal/include/algorithms/engines/mrg32k3a/mrg32k3a.h>
+#include <daal/include/algorithms/engines/philox4x32x10/philox4x32x10.h>
+#include <daal/include/algorithms/engines/mt19937/mt19937.h>
+
+#include "oneapi/dal/backend/primitives/rng/utils.hpp"
+#include "oneapi/dal/backend/primitives/rng/rng_types.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+/// A class that provides an interface for random number generation on the host (CPU) only.
+///
+/// This class serves as a wrapper for host-based random number generators (RNGs), supporting multiple engine
+/// types for flexible and efficient random number generation on CPU. It abstracts the underlying engine
+/// implementation and provides an interface to manage and retrieve the engine's state.
+///
+/// @tparam EngineType The RNG engine type to be used. Defaults to `engine_type::mt2203`.
+///
+/// @param[in] seed  The initial seed for the random number generator. Defaults to `777`.
+///
+/// @note The class only supports host-based RNG and does not require a SYCL queue or device context.
+class host_engine {
+public:
+    explicit host_engine(std::int64_t seed = 777, engine_type method = engine_type::mt2203) {
+        switch (method) {
+            case engine_type::mt2203:
+                host_engine_ = daal::algorithms::engines::mt2203::Batch<>::create(seed);
+                break;
+            case engine_type::mcg59:
+                host_engine_ = daal::algorithms::engines::mcg59::Batch<>::create(seed);
+                break;
+            case engine_type::mrg32k3a:
+                host_engine_ = daal::algorithms::engines::mrg32k3a::Batch<>::create(seed);
+                break;
+            case engine_type::philox4x32x10:
+                host_engine_ = daal::algorithms::engines::philox4x32x10::Batch<>::create(seed);
+                break;
+            case engine_type::mt19937:
+                host_engine_ = daal::algorithms::engines::mt19937::Batch<>::create(seed);
+                break;
+            default: throw std::invalid_argument("Unsupported engine type 1");
+        }
+        impl_ =
+            dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(host_engine_.get());
+        if (!impl_) {
+            throw std::domain_error("RNG engine is not supported");
+        }
+    }
+
+    explicit host_engine(const daal::algorithms::engines::EnginePtr& eng) : host_engine_(eng) {
```

Comment:
I do not think that this is a proper fix to rule of three. Because a copy constructor should accept the object of the same type `host_engine`.